hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/token/ShortLivedTokenVerifier.java

```diff
@@ -0,0 +1,135 @@
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hdds.security.token;
+
+import org.apache.hadoop.hdds.protocol.datanode.proto.ContainerProtos;
+import org.apache.hadoop.hdds.protocol.datanode.proto.ContainerProtos.ContainerCommandRequestProtoOrBuilder;
+import org.apache.hadoop.hdds.security.exception.SCMSecurityException;
+import org.apache.hadoop.hdds.security.x509.SecurityConfig;
+import org.apache.hadoop.hdds.security.x509.certificate.client.CertificateClient;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.security.token.Token;
+
+import java.io.ByteArrayInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.security.cert.CertificateExpiredException;
+import java.security.cert.CertificateNotYetValidException;
+import java.security.cert.X509Certificate;
+import java.time.Instant;
+import java.util.Objects;
+
+/**
+ * Verifies short-lived token.
+ * @param <T> type of short-lived token
+ */
+public abstract class
+    ShortLivedTokenVerifier<T extends ShortLivedTokenIdentifier>
+    implements TokenVerifier {
+
+  private final CertificateClient caClient;
+  private final SecurityConfig conf;
+
+  protected ShortLivedTokenVerifier(SecurityConfig conf,
+      CertificateClient caClient) {
+    this.conf = conf;
+    this.caClient = caClient;
+  }
+
+  /** Whether the specific kind of token is required for {@code cmdType}. */
+  protected abstract boolean isTokenRequired(ContainerProtos.Type cmdType);
+
+  /** Create new empty token identifier, to be filled from Token. */
+  protected abstract T createTokenIdentifier();
+
+  /** Extract info on "service" being accessed by {@code cmd}. */
+  protected abstract Object getService(
+      ContainerCommandRequestProtoOrBuilder cmd);
+
+  /** Hook for further verification. */
+  protected void verify(T tokenId, ContainerCommandRequestProtoOrBuilder cmd)
+      throws SCMSecurityException {
+    // NOP
+  }
+
+  @Override
+  public void verify(String user, Token<?> token,
+      ContainerCommandRequestProtoOrBuilder cmd) throws SCMSecurityException {
+
+    if (!isTokenRequired(cmd.getCmdType())) {
+      return;
+    }
+
+    if (caClient == null) {
+      throw new SCMSecurityException("Certificate client not available " +
+          "to validate token");
+    }
+
+    T tokenId = createTokenIdentifier();
+    try {
+      tokenId.readFields(new DataInputStream(new ByteArrayInputStream(
+          token.getIdentifier())));
+    } catch (IOException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 103, 106.